src/common/tracer.h

```diff
@@ -0,0 +1,80 @@
+#ifndef TRACER_H_
+#define TRACER_H_
+
+#define SIGNED_RIGHT_SHIFT_IS 1
+#define ARITHMETIC_RIGHT_SHIFT 1
+
+#include<cstring>
+#include <arpa/inet.h>
+#include <yaml-cpp/yaml.h>
+#include<atomic>
+
+#include <jaegertracing/Tracer.h>
+#include <jaegertracing/net/IPAddress.h>
+#include <jaegertracing/net/Socket.h>
+
+
+typedef std::unique_ptr<opentracing::Span> Span;
+
+struct Jaeger_Tracer{
+    Jaeger_Tracer(){}
+    ~Jaeger_Tracer();
+  std::shared_ptr<opentracing::v3::Tracer> tracer = NULL;
+};
+
+//will be used throughout ceph to create spans
+extern Jaeger_Tracer tracer;
+extern std::atomic<bool> jaeger_initialized;
+
+
+static inline void init_tracer(const char* tracerName,const char* filePath){
+      if(jaeger_initialized) return;
+
+      try{
+          auto yaml = YAML::LoadFile(filePath);
+          auto configuration = jaegertracing::Config::parse(yaml);
+
+          jaegertracing::net::Socket socket;
+          socket.open(AF_INET, SOCK_STREAM);
+          const std::string serverURL = configuration.sampler().samplingServerURL();
+          socket.connect(serverURL); // this is used to check if the tracer is able to connect with server successfully
+
+          tracer.tracer = jaegertracing::Tracer::make(
+          tracerName,
+          configuration,
+          jaegertracing::logging::consoleLogger());
+      }catch(...) {return;}
+      opentracing::Tracer::InitGlobal(
+      std::static_pointer_cast<opentracing::Tracer>(tracer.tracer));
+      jaeger_initialized = true;
```

Comment:
this does not protect you from multiple initializations.

Reply:
it does at the start of the function there is a atomic variable to check if it been initialized before
`if(jaeger_initialized) return;`

Reply:
but race condition can still happen. you set the variable after you actually initialize

Reply:
By initialied before I means it checks if value is true or not. Do you mean I m assigning value to it before intializing it?

Reply:
i mean that if 2 threads are going to call this function, there is a chance that both will initialize before one of them sets the atomic boolean